From 1029371062ea52453fa2770e93ca7a4ae532b134 Mon Sep 17 00:00:00 2001
From: Mao <m@maowtm.org>
Date: Sat, 4 Jan 2025 00:04:25 +0000
Subject: [PATCH 2/4] ptrace_traceme return 0

---
 kernel/ptrace.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/kernel/ptrace.c b/kernel/ptrace.c
index d5f89f9ef29f..017d25c00fa3 100644
--- a/kernel/ptrace.c
+++ b/kernel/ptrace.c
@@ -486,7 +486,7 @@ static int ptrace_attach(struct task_struct *task, long request,
  */
 static int ptrace_traceme(void)
 {
-	int ret = -EPERM;
+	int ret = 0;
 
 	write_lock_irq(&tasklist_lock);
 	/* Are we already being traced? */
-- 
2.47.1

