From a5480e257defecd557c1e3b82382acd94e230a8b Mon Sep 17 00:00:00 2001
From: Tingmao Wang <m@maowtm.org>
Date: Mon, 14 Apr 2025 23:47:36 +0100
Subject: [PATCH 17/18] restart_syscall

---
 fs/read_write.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/fs/read_write.c b/fs/read_write.c
index 17e6ab568c02..45c0a8a80ebc 100644
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -790,8 +790,7 @@ ssize_t ksys_write(unsigned int fd, const char __user *buf, size_t count)
 				return ret;
 			}
 
-			/* Hmm... what do we do now? Let's just return for now. */
-			return count;
+			return restart_syscall();
 		} else {
 			pr_info("hack: %u gave different output\n", current->hack_target.next_number - 1);
 			/*
-- 
2.39.5

